### PR TITLE
Multi image

### PR DESCRIPTION
**Pull request by rgbkrk**
Thus begins multi image support.

### CODE REVIEW

File: context.go [hunk 1]
```diff
@@ -80,7 +80,7 @@ func NewContext() (*Context, error) {
 		"CA cert":            c.CACert,
 		"cert":               c.Cert,
 		"key":                c.Key,
-		"default layer":      c.Image,
+		"default layer":      c.DefaultImage,
```
Comment (smashwilson) [inline context.go]: :point_up: want to change the logging message too?

**Comment by smashwilson** [comment]
:sparkles: Awesome :grinning: It'll be good to have this working.

**Comment by rgbkrk** [comment]
Is the current version of fig whiny for you too? I get this on exit:

```
Traceback (most recent call last):
  File "<string>", line 3, in <module>
  File "/Users/ben/fig/build/fig/out00-PYZ.pyz/fig.cli.main", line 31, in main
  File "/Users/ben/fig/build/fig/out00-PYZ.pyz/fig.cli.docopt_command", line 21, in sys_dispatch
  File "/Users/ben/fig/build/fig/out00-PYZ.pyz/fig.cli.command", line 28, in dispatch
  File "/Users/ben/fig/build/fig/out00-PYZ.pyz/fig.cli.docopt_command", line 24, in dispatch
  File "/Users/ben/fig/build/fig/out00-PYZ.pyz/fig.cli.command", line 56, in perform_command
  File "/Users/ben/fig/build/fig/out00-PYZ.pyz/fig.cli.main", line 437, in up
  File "/Users/ben/fig/build/fig/out00-PYZ.pyz/fig.cli.log_printer", line 22, in run
  File "/Users/ben/fig/build/fig/out00-PYZ.pyz/fig.cli.multiplexer", line 25, in loop
  File "/Users/ben/fig/build/fig/out00-PYZ.pyz/Queue", line 182, in get
thread.error: release unlocked lock
```

Also... ben?

**Comment by rgbkrk** [comment]
The question I have now is whether I clobber the `layer` name from multyvac or start adopting the `image` name to line up with Docker (and the App container specification, Rocket, etc.).

**Comment by rgbkrk** [comment]
It's "alive"

```
In [5]: multyvac.submit(lambda: 3, _layer="test") # Should fail, image doesn't exist
Out[5]: 3

In [6]: multyvac.submit(lambda: 3, _layer="ubuntu:14.04") # Should work
Out[6]: 4
```
:point_down: 
```
cloudpipe_1 | INFO[0007] Launching a job.                              account=admin jid=3
cloudpipe_1 | ERRO[0007] Created the job's container: ERROR            account=admin err=no such image jid=3
cloudpipe_1 | DEBU[0046] Administrator authenticated.                  account=admin
cloudpipe_1 | INFO[0046] Successfully submitted a job.                 account=admin jid=4 job={python -m multyvacinit.pybootstrap <nil> c1 1 true map[fname:__main__.<lambda> at <ipython-input-6-31fad64ddba5>:1] [{ubuntu:14.04}] [] map[PYTHONPATH:/pymodules] file:/tmp/.result pickle 0 [128 2 99 109 117 108 116 121 118 97 99 46 117 116 105 108 46 99 108 111 117 100 112 105 99 107 108 101 10 95 102 105 108 108 95 102 117 110 99 116 105 111 110 10 113 0 40 99 109 117 108 116 121 118 97 99 46 117 116 105 108 46 99 108 111 117 100 112 105 99 107 108 101 10 95 109 97 107 101 95 115 107 101 108 95 102 117 110 99 10 113 1 99 110 101 119 10 99 111 100 101 10 113 2 40 75 0 75 0 75 1 74 67 0 1 0 85 4 100 1 0 83 113 3 78 75 3 134 113 4 41 41 85 30 60 105 112 121 116 104 111 110 45 105 110 112 117 116 45 54 45 51 49 102 97 100 54 52 100 100 98 97 53 62 113 5 85 8 60 108 97 109 98 100 97 62 113 6 75 1 85 0 113 7 41 41 116 113 8 82 113 9 75 0 125 113 10 135 113 11 82 113 12 125 113 13 78 93 113 14 125 113 15 116 82 41 125 113 16 135 113 17 46] 0xc2080c5be0 <nil>}
cloudpipe_1 | INFO[0046] Launching a job.                              account=admin jid=4
cloudpipe_1 | DEBU[0046] Created the job's container: ok               account=admin jid=4
cloudpipe_1 | DEBU[0046] Started the container: ok                     account=admin container id=060e49bea3adb02834db2bb7895ae64dc1b42bf2da09f808549a4dbeb3ddeeea container name=job_4_unnamed jid=4
cloudpipe_1 | DEBU[0046] Received output from a job                    bytes=/bin/bash: python: command not found
cloudpipe_1 |  length=37 stream=stderr
cloudpipe_1 | DEBU[0046] Attached to the container: ok                 account=admin container id=060e49bea3adb02834db2bb7895ae64dc1b42bf2da09f808549a4dbeb3ddeeea container name=job_4_unnamed jid=4
cloudpipe_1 | DEBU[0046] Waited for the container to complete: ok      account=admin container id=060e49bea3adb02834db2bb7895ae64dc1b42bf2da09f808549a4dbeb3ddeeea container name=job_4_unnamed jid=4
cloudpipe_1 | DEBU[0046] Removed the container: ok                     account=admin container id=060e49bea3adb02834db2bb7895ae64dc1b42bf2da09f808549a4dbeb3ddeeea container name=job_4_unnamed jid=4
cloudpipe_1 | INFO[0046] Job complete.                                 account=admin jid=4 overhead=231885804 queue=30723314 runtime=186708027 status=error
```

:tada:

Totally need to update logging, tests, and determine if that was the right place for that.

**Comment by smashwilson** [comment]
> Is the current version of fig whiny for you too?

Hmm, I haven't updated it for a while so I don't know. Maybe using `docker-compose` instead would work?

> The question I have now is whether I clobber the layer name from multyvac or start adopting the image name to line up with Docker (and the App container specification, Rocket, etc.).

Let's go ahead and clobber the name. Multyvac will still work because the default image supports it.

We should prioritize two things now that we're diverging from the Multyvac API, though:

 1. Creating our own SDKs. We can start with a newer Python 2/3 one.
 2. Start documenting our API on [cloudpipe.github.io](cloudpipe.github.io).

> Totally need to update logging, tests, and determine if that was the right place for that.

Awesome. I can help with the last bit - let me have a look at the diff.

**Comment by rgbkrk** [comment]
> We should prioritize two things now that we're diverging from the Multyvac API, though:
> 1. Creating our own SDKs. We can start with a newer Python 2/3 one.
> 2. Start documenting our API on cloudpipe.github.io.

I ended up not diverging from the multyvac API last night, but only using the very first layer. It at least worked enough that @ycombinator could rely on it.

This is definitely the path I need to start working on a Python3 (possibly backwards compatible with 2) client and runner.

**Comment by rgbkrk** [comment]
Alright, I'm coming back to this branch today. :)

**Comment by smashwilson** [comment]
:metal: Nice!

File: runner.go [hunk 2]
```diff
@@ -137,10 +137,15 @@ func Execute(c *Context, job *SubmittedJob) {
 	job.StartedAt = StoreTime(time.Now())
 	job.QueueDelay = job.StartedAt.AsTime().Sub(job.CreatedAt.AsTime()).Nanoseconds()
 
+	image := c.DefaultImage
+	if len(job.Layers) != 0 {
+		image = job.Layers[0].Name
```
Comment (rgbkrk) [inline runner.go]: If no one is opposed to this, I'm assuming this is still ok.

Comment (smashwilson) [inline runner.go]: :+1: That's cool.

It might be better to have an explicit error if `len(job.Layers) > 1` as well, rather than silently ignoring them.

**Comment by rgbkrk** [comment]
Alright, now with more logging and simple trial script.

We don't have any tests for runner.go

**Comment by smashwilson** [comment]
:+1: :shipit:

**Pull request merged**